Тетрадка предназначена для студентов образвательного проекта Booster MDS использование вне проекта запрещено.

##### Задания не несут смысловой нагрузки для реальной жизни, они сделаны лишь для того, чтоб научить использовать Pandas

#### Каждое задание оценивает в 1 балл, для успешной сдачи задания нужно набрать 12 баллов

This database contains 76 attributes, but all published experiments refer to using a subset of 14 of them. In particular, the Cleveland database is the only one that has been used by ML researchers to this date. The "goal" field refers to the presence of heart disease in the patient. It is integer valued from 0 (no presence) to 1.

https://www.kaggle.com/ronitf/heart-disease-uci

___

In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv('E:\Programming\Booster MDS\heart_pandas.csv')

In [3]:
data.head(5)

age     sex   chest_pain_type  resting_blood_pressure  cholesterol  \
0   63    male  non-anginal pain                     145          233   
1   37    male   atypical angina                     130          250   
2   41  female    typical angina                     130          204   
3   56    male    typical angina                     120          236   
4   57  female    typical angina                     120          354   

     fasting_blood_sugar               rest_ecg  max_heart_rate_achieved  \
0  greater than 120mg/ml                 normal                      150   
1    lower than 120mg/ml  ST-T wave abnormality                      187   
2    lower than 120mg/ml                 normal                      172   
3    lower than 120mg/ml  ST-T wave abnormality                      178   
4    lower than 120mg/ml  ST-T wave abnormality                      163   

  exercise_induced_angina  st_depression   st_slope  num_major_vessels  \
0                      no            2.3  upsloping                  0   
1                      no            3.5  upsloping                  0   
2                      no            1.4       flat                  0   
3                      no            0.8       flat                  0   
4                     yes            0.6       flat                  0   

    thalassemia  target  
0        normal       1  
1  fixed defect       1  
2  fixed defect       1  
3  fixed defect       1  
4  fixed defect       1

#### Описание признаков

**age**: The person's age in years

**sex**: The person's sex (1 = male, 0 = female)

**cp**: The chest pain experienced (Value 1: typical angina, Value 2: atypical angina, Value 3: non-anginal pain, Value 4: asymptomatic)

**trestbps**: The person's resting blood pressure (mm Hg on admission to the hospital)

**chol**: The person's cholesterol measurement in mg/dl

**fbs**: The person's fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)

**restecg**: Resting electrocardiographic measurement (0 = normal, 1 = having ST-T wave abnormality, 2 = showing probable or definite left ventricular hypertrophy by Estes' criteria)

**thalach**: The person's maximum heart rate achieved

**exang**: Exercise induced angina (1 = yes; 0 = no)

**oldpeak**: ST depression induced by exercise relative to rest ('ST' relates to positions on the ECG plot. See more here)

**slope**: the slope of the peak exercise ST segment (Value 1: upsloping, Value 2: flat, Value 3: downsloping)

**ca**: The number of major vessels (0-3)

**thal**: A blood disorder called thalassemia (3 = normal; 6 = fixed defect; 7 = reversable defect)

**target**: Heart disease (0 = no, 1 = yes)

> 1. age 
> 2. sex 
> 3. chest pain type (4 values) 
> 4. resting blood pressure 
> 5. serum cholestoral in mg/dl 
> 6. fasting blood sugar > 120 mg/dl
> 7. resting electrocardiographic results (values 0,1,2)
> 8. maximum heart rate achieved 
> 9. exercise induced angina 
> 10. oldpeak = ST depression induced by exercise relative to rest 
> 11. the slope of the peak exercise ST segment 
> 12. number of major vessels (0-3) colored by flourosopy 
> 13. thal: 3 = normal; 6 = fixed defect; 7 = reversable defect

___

**Сколько мужчин в датасете? Сколько женщен? (sex) **

In [ ]:
## Ваш код
data["sex"].value_counts()

In [ ]:
## Ваш ответ:
Мужчин 207, женщин 96

 **Какой процент мужчин в датасете? (решите в одну строчку, не используя предыдущий результат. Не забудте знак процента) **

In [ ]:
## Ваш код
print("{}%".format(sum(data["sex"] == "male" ) / len(data) * 100))

In [ ]:
## Ваш ответ:
68%

**Сколько мужчин имеют заболевание сердца? Сколько женщин имеют заболевание сердца? **

In [ ]:
## Ваш код
print(sum((data['target']==1) & (data['sex']=='male')))
print(sum((data['target']==1) & (data['sex']=='female')))

In [ ]:
## Ваш ответ:
мужчины: 93
Женщины: 72

**Какую долю, от общего числа пациентов, занимают мужчины не имеющие болезнь сердца?**

In [ ]:
## Ваш код
((sum((data["target"] == 0) & (data["sex"] == "male"))) /len(data)) * 100

In [ ]:
## Ваш ответ:
37,6

**Сколько лет самому молодому пациенту, страдающему болезнью сердца?**

In [ ]:
## Ваш код
data[data["target"] == 1]["age"].min() 

In [ ]:
## Ваш ответ:
29 лет

**Сколько лет самому возастному пациенту, у которого нет проблем с сердцем?**

In [ ]:
## Ваш код
data[data["target"] == 0]["age"].max()

In [ ]:
## Ваш ответ:
77 лет

**Сколько лет самой молодой женщине, которая страдает болезнью сердца?**

In [ ]:
## Ваш код
data[(data["target"] == 1) & (data["sex"] == "female")]["age"].min()

In [ ]:
## Ваш ответ:
34

**Какой средний возраст женщин?**

In [ ]:
## Ваш код
data[data["sex"] == "female"]["age"].mean()

In [ ]:
## Ваш ответ:
55,7

**Каковы средние значения и среднеквадратичные отклонения возраста тех, кто страдают болезнью сердца?**

In [ ]:
## Ваш код
print(data[data["target"] == 1]["age"].mean()) # среднее значение
print(data[data["target"] == 1]["age"].std()) #среднеквадратичное отклонение

In [ ]:
## Ваш ответ:
Среднее отклонение 52.5
Среднеквадратичное 9,6

** Правда ли, что люди не болеющие болезнью сердца имеют уровня холестерина меньше среднего? (chol) **

In [ ]:
## Ваш код
print(data[data["target"] == 0]["cholesterol"].mean())
print(data["cholesterol"].mean())
print(data[data["target"] == 0]["cholesterol"].mean() < data["cholesterol"].mean())

In [ ]:
## Ваш ответ:
Ффранье!

**Выведите статистику rest_ecg для все числовых признаков, его максимально и среднее значение (используйте groupby(), решите в одну строчку)**

In [6]:
## Ваш код
data.groupby(by = 'rest_ecg').agg([np.max, np.mean])

age            resting_blood_pressure  \
                             amax       mean                   amax   
rest_ecg                                                              
ST-T wave abnormality          71  52.914474                    180   
left ventricular hypertrophy   76  61.000000                    180   
normal                         77  55.687075                    200   

                                         cholesterol              \
                                    mean        amax        mean   
rest_ecg                                                           
ST-T wave abnormality         129.065789         354  237.269737   
left ventricular hypertrophy  140.500000         327  261.750000   
normal                        134.027211         564  255.142857   

                             max_heart_rate_achieved              \
                                                amax        mean   
rest_ecg                                                           
ST-T wave abnormality                            194  151.960526   
left ventricular hypertrophy                     140  125.750000   
normal                                           202  147.904762   

                             st_depression           num_major_vessels  \
                                      amax      mean              amax   
rest_ecg                                                                 
ST-T wave abnormality                  5.6  0.879605                 4   
left ventricular hypertrophy           4.4  2.725000                 3   
normal                                 6.2  1.159184                 4   

                                       target            
                                  mean   amax      mean  
rest_ecg                                                 
ST-T wave abnormality         0.638158      1  0.631579  
left ventricular hypertrophy  1.000000      1  0.250000  
normal                        0.816327      1  0.462585

In [ ]:
## Ваш ответ:
не знаю, что сюда писать. Смотри ответ выше

**Посчитайте у кого уровень депрессии при физический нагрузке выше (в среднем), среди мужчин страдающих болезнью сердца или среди женщин не страдающих болезнью сердца (st_depression)**

In [ ]:
## Ваш код
a = data[(data["target"] == 1) & (data["sex"] == "male")]["st_depression"].mean()
b = data[(data["target"] == 0) & (data["sex"] == "female")]["st_depression"].mean()
if a > b:
    print("Male")
elif a < b:
    print("Female")
else:
    print("Equal")

In [ ]:
## Ваш ответ:
Женщины. Не думаю, что код оптимален, зато понятно! В отличии от той фигни, что я дальше пишу

**Посчтитайте максимальный и минимальный уровень холестерина для каждого типа chest_pain_type, rest_ecg, thalassemia. Пишите код оптимально, можно использовать циклы**

In [28]:
## Ваш код
result = pd.DataFrame()
for c in 'chest_pain_type',"rest_ecg",'thalassemia':
    for t in data[c].unique():
        minmax = {"min": data[data[c] == t]['cholesterol'].min(), "max": data[data[c] == t]['cholesterol'].max(), "c": c, "t": t}
        result = result.append(minmax, ignore_index=True)
result
        

c    max    min                             t
0  chest_pain_type  298.0  182.0              non-anginal pain
1  chest_pain_type  564.0  126.0               atypical angina
2  chest_pain_type  409.0  131.0                typical angina
3         rest_ecg  564.0  149.0                        normal
4         rest_ecg  354.0  126.0         ST-T wave abnormality
5         rest_ecg  327.0  197.0  left ventricular hypertrophy
6      thalassemia  318.0  169.0                        normal
7      thalassemia  417.0  141.0                  fixed defect
8      thalassemia  564.0  126.0             reversable defect

In [29]:
## Ваш ответ:
result

c    max    min                             t
0  chest_pain_type  298.0  182.0              non-anginal pain
1  chest_pain_type  564.0  126.0               atypical angina
2  chest_pain_type  409.0  131.0                typical angina
3         rest_ecg  564.0  149.0                        normal
4         rest_ecg  354.0  126.0         ST-T wave abnormality
5         rest_ecg  327.0  197.0  left ventricular hypertrophy
6      thalassemia  318.0  169.0                        normal
7      thalassemia  417.0  141.0                  fixed defect
8      thalassemia  564.0  126.0             reversable defect

**Сколько значений может принимать каждый из категориальных признаков?**

In [9]:
## Ваш код
categorical_columns = [c for c in data.columns if data[c].dtype.name == 'object']
for t in categorical_columns:
    print(t, data[t].nunique())

sex 2
chest_pain_type 3
fasting_blood_sugar 2
rest_ecg 3
exercise_induced_angina 2
st_slope 2
thalassemia 3


In [ ]:
## Ваш ответ:
sex 2
chest_pain_type 3
fasting_blood_sugar 2
rest_ecg 3
exercise_induced_angina 2
st_slope 2
thalassemia 3

**У какого категориального признака наблюдается самый сильный дизбаланс классов?**

In [10]:
## Ваш код
def disbalance(f):
    return min(data[f].value_counts())/max(data[f].value_counts())

pd.Series(list(map(disbalance,categorical_columns)),index=categorical_columns).sort_values().head(1).index[0]

'rest_ecg'

In [ ]:
## Ваш ответ:
'rest_ecg'

---

@Booster
Мамкин Data Scientist

tg @mommyscience